crates/pallet-native-to-evm-swap/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+            evm_transaction_hash: H256,
+        },
+    }
+
+    #[pallet::call(weight(<T as Config>::WeightInfo))]
+    impl<T: Config> Pallet<T> {
+        /// Swap balances.
+        #[pallet::call_index(0)]
+        pub fn swap(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Expendable)?;
+
+                Ok(())
+            })
+        }
+
+        /// Same as the swap call, but with a check that the swap will not kill the origin account.
+        #[pallet::call_index(1)]
+        pub fn swap_keep_alive(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Preserve)?;
+
+                Ok(())
+            })
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// General swap balances implementation.
+        pub fn do_swap(
+            who: T::AccountId,
+            to: T::EvmAccountId,
+            amount: NativeBalanceOf<T>,
+            preservation: Preservation,
+        ) -> DispatchResult {
+            let estimated_swapped_balance = T::BalanceConverter::convert(amount);
+            T::EvmToken::can_deposit(&to, estimated_swapped_balance, Provenance::Extant)
+                .into_result()?;
+
+            T::NativeToken::transfer(&who, &T::BridgePotNative::get(), amount, preservation)?;
+
+            let evm_balance_to_be_deposited: u128 =
+                estimated_swapped_balance.unique_saturated_into();
+
+            let transaction = pallet_ethereum::Transaction::EIP1559(ethereum::EIP1559Transaction {
+                chain_id: <T as pallet_evm::Config>::ChainId::get(),
+                nonce: pallet_evm::Pallet::<T>::account_basic(&T::BridgePotEvm::get().into())
+                    .0
+                    .nonce,
+                max_priority_fee_per_gas: 0.into(),
+                max_fee_per_gas: 0.into(),
+                gas_limit: <T as pallet_evm::Config>::config()
+                    .gas_transaction_call
+                    .into(), // simple transfer
+                action: ethereum::TransactionAction::Call(to.clone().into()),
+                value: U256::from(evm_balance_to_be_deposited),
+                input: Default::default(),
+                access_list: Default::default(),
+                odd_y_parity: false,
+                r: Default::default(),
+                s: Default::default(),
+            });
+
+            let evm_transaction_hash = transaction.hash();
+
+            let _post_info = pallet_ethereum::ValidatedTransaction::<T>::apply(
```

Comment:
To revisit the `PostDispatchInfo` question; turns out (after inspection of the `pallet_ethereum::ValidatedTransaction` implementation) the `post_info` is highly non-trivial; it has `pays: No`, weight based on the EVM gas consumed (which is, nonetheless, covered by benches if we bench this call).

So, maybe we should, at least, apply some scrutiny to the returned value, instead of just `_post_info`. A `debug_assert!` would be a good idea here.

Reply:
That's why I asked before, good suggestion that we can add `debug_assert` to verify that we consume expected gas, added - https://github.com/humanode-network/humanode/pull/1449/commits/067f1c39f22811decc858ab472a064c1f474494a

Reply:
Moved ethereum transfer logic to separate fn - https://github.com/humanode-network/humanode/pull/1449/commits/eaf3920eb2ab819020d6cf7ab377222d4b1c4c72